pkg/kstatus/observe/reader/cached_reader.go

```diff
@@ -0,0 +1,205 @@
+// finding all combinations of GroupVersionKind and namespace referenced by the provided
+// identifiers. This list is then expanded to include any known generated resource types.
+type CachedObserverReader struct {
+	sync.RWMutex
+
+	reader client.Reader
+
+	mapper meta.RESTMapper
+
+	gns []gvkNamespace
+
+	cache map[gvkNamespace]unstructured.UnstructuredList
+}
+
+// gvkNamespace contains information about a GroupVersionKind and a namespace.
+type gvkNamespace struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+}
+
+// Get looks up the resource identified by the key and the object GVK in the cache. If the needed combination
+// of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachedObserverReader) Get(_ context.Context, key client.ObjectKey, obj *unstructured.Unstructured) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	mapping, err := c.mapper.RESTMapping(gvk.GroupKind())
+	if err != nil {
+		return err
+	}
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: key.Namespace,
+	}
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+	for _, u := range cacheList.Items {
+		if u.GetName() == key.Name {
+			obj.Object = u.Object
+			return nil
+		}
+	}
+	return errors.NewNotFound(mapping.Resource.GroupResource(), key.Name)
+}
+
+// ListNamespaceScoped lists all resource identifier by the GVK of the list, the namespace and the selector
+// from the cache. If the needed combination of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachedObserverReader) ListNamespaceScoped(_ context.Context, list *unstructured.UnstructuredList, namespace string, selector labels.Selector) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := list.GroupVersionKind()
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: namespace,
+	}
+
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+
+	var items []unstructured.Unstructured
+	for _, u := range cacheList.Items {
+		if selector.Matches(labels.Set(u.GetLabels())) {
+			items = append(items, u)
+		}
+	}
+	list.Items = items
+	return nil
+}
+
+// ListNamespaceScoped lists all resource identifier by the GVK of the list and selector
+// from the cache. If the needed combination of GVK and namespace is not part of the cache,
+// that is considered an error.
+func (c *CachedObserverReader) ListClusterScoped(ctx context.Context, list *unstructured.UnstructuredList, selector labels.Selector) error {
+	return c.ListNamespaceScoped(ctx, list, "", selector)
+}
+
+// Sync loops over the list of gvkNamespace we know of, and uses list calls to fetch the resources.
+// This information populates the cache.
+func (c *CachedObserverReader) Sync(ctx context.Context) error {
+	c.Lock()
```

Comment:
Each call to `StatusObserver.Observe` will create a new `StatusObserverRunner` that will run in its own go routine. So with there is no way `Sync` can be called concurrently since both the `ObserverReader` and the observers are local to the `StatusObserverRunner`. Everything that happens in the runner is confined to a single goroutine.

Currently all the observers are also running in that same goroutine, so we don't strictly need any synchronization here at all. I am thinking about running multiple observers in parallell when using the cached `ObserverReader`, but I haven't gotten to it yet. It would involve some kind of task queue and a limited number of goroutines to avoid creating too many of them. Either way, the plan is that `Sync` will only ever be called by the main goroutine for the runner.

If we are not using the cached `ObserverReader`, running many observers in parallell just means more load on the APIServer, so it doesn't seem like such a good idea.